Content creation for multi-instance Argo CD feature (#89312)
openshift/openshift-docs: modules/con_enabling_annotation_based_resource_tracking_in_argo_cd.adoc

* If multiple instances have applications with the same name, set resource tracking to `annotation+label`.
* If issues arise, check the *Argo CD Application* status and logs in the {OCP} Web Console.
====

.Pre-requisites

* You have logged in to the {OCP} cluster as an administrator.
* You have installed the {gitops-title} Operator on your {OCP} cluster.

[id="configuring-annotation-based-tracking-in-multiple-argo-cd-instances_{context}"]
== Configuring annotation-based tracking in multiple Argo CD instances

You can configure annotation-based tracking in multiple Argo CD instances by completing the following steps.

. Log in to the {OCP} web console. 

. In the *Administrator* perspective of the web console, click *Operators* -> *Installed Operators*.

. Create or select the project where you want to install the user-defined Argo CD instance from the *Project* list.

. Select *{gitops-title}* from the installed Operators list and click the *Argo CD* tab.

. Click *Create ArgoCD* to create two Argo CD instances using the following YAML snippet.
+
.Example Argo CD instance with an annotation label
[source,yaml]
----

> apiVersion: argoproj.io/v1beta1
> kind: ArgoCD
> metadata:
>   name: argocd-instance-demo-1 #<1>
>   namespace: argocd-test-demo-1 #<2>
>   name: argocd-instance-demo-2 #<3>
>   namespace: argocd-test-demo-2 #<4>
> spec:
>   installationID: "instance-demo-1" #<5>
>   resourceTrackingMethod: "annotation+label"
>   installationID: "instance-demo-2" #<6>
>   resourceTrackingMethod: "annotation+label"

Reviewer: Its not possible to create an instance this way as you will get `duplicate key` error.

This is valid:

```suggestion
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: argocd-instance-demo-1 #<1>
  namespace: argocd-test-demo-1 #<2>
spec:
  installationID: "instance-demo-1" #<3>
  resourceTrackingMethod: "annotation+label"
```

Author: Fixed.

> ----
> apiVersion: argoproj.io/v1beta1
> kind: ArgoCD
> metadata:
>   name: argocd-instance-demo-1 # <1>
>   namespace: argocd-test-demo-1 # <2>
>   name: argocd-instance-demo-2 # <3>
>   namespace: argocd-test-demo-2 # <4>
> spec:
>   installationID: "instance-demo-1" # <5>
>   resourceTrackingMethod: "annotation+label"
>   installationID: "instance-demo-2" # <6>
>   resourceTrackingMethod: "annotation+label"
> ----

Reviewer: ```suggestion
----
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: argocd-instance-demo-1 # <1>
  namespace: argocd-test-demo-1 # <2>
spec:
  installationID: "instance-demo-1" # <3>
  resourceTrackingMethod: "annotation+label"
----
----
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: argocd-instance-demo-2 # <4>
  namespace: argocd-test-demo-2 # <5>
spec:
  installationID: "instance-demo-2" # <6>
  resourceTrackingMethod: "annotation+label"
----
```
Change the explanations for <1> to <6> accordingly

Author: Fixed.

+
<1> Specifies the name of the first Argo CD instance.
<2> Specifies the namespace used for the first Argo CD instance.
<3> Specifies the name of the second Argo CD instance.
<4> Specifies the namespace used for the second Argo CD instance.
<5> Specifies the name of the `installationID` for the first Argo CD instance.
<6> Specifies the name of the `installationID` for the second Argo CD instance.
+
. Configure and label target namespaces to associate namespaces with their Argo CD instances.
.. Navigate to *Administration* -> *Namespaces*.
.. Create namespaces for application deployments, `app-ns-1` and `app-ns-2`.
.. To associate each namespace with their respective Argo CD instances, run the following command:
+
[source,terminal]
----
oc label namespace app-ns-1 argocd.argoproj.io/managed-by=argocd-test-1

> oc label namespace app-ns-2 argocd.argoproj.io/managed-by=argocd-test-2

Author: Fixed.

----
+
. Create two applications in Argo CD.
.. In the OpenShift Web Console, go to *Operators* → *Installed Operators* → *OpenShift GitOps Operator*.
.. Select *Argo CD* and navigate to the *Applications* tab.
.. Click *Create Application*.
.. Enter the following YAML snippet to create two applications in Argo CD.
+
.Example application creation using Argo CD
[source,yaml]
----
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: sprint-petclinic # <1>
  namespace: argocd-test-1 # <2>

>   name: sprint-petclinic # <3>

Reviewer 2: Can't club 2 applications like this. This will cause error. Instead have 2 seperate `Application` objects.

Author: Fixed.

>   namespace: argocd-test-2 #<4>

Reviewer: Avoid duplicate keys

Author: Fixed.

> ----
> apiVersion: argoproj.io/v1alpha1
> kind: Application
> metadata:
>   name: sprint-petclinic # <1>
>   namespace: argocd-test-1 # <2>
>   name: sprint-petclinic # <3>
>   namespace: argocd-test-2 # <4>
> spec:
>   project: default
>   source:
>     repoURL: https://github.com/redhat-developer/gitops-operator
>     path: test/examples/nginx
>     targetRevision: HEAD
>   destination:
>     server: https://kubernetes.default.svc
>     namespace: app-ns-1
>   syncPolicy:
>     automated: {}
>   project: default
>   source:
>     repoURL: https://github.com/redhat-developer/gitops-operator
>     path: test/examples/nginx
>     targetRevision: HEAD
>   destination:
>     server: https://kubernetes.default.svc
>     namespace: app-ns-2
>   syncPolicy:
>     automated: {}
> ----

Reviewer: ```suggestion
----
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: sprint-petclinic 
  namespace: argocd-test-demo-1
spec:
  project: default
  source:
    repoURL: https://github.com/redhat-developer/gitops-operator
    path: test/examples/nginx
    targetRevision: HEAD
  destination:
    server: https://kubernetes.default.svc
    namespace: app-ns-1
  syncPolicy:
    automated: {}
----
----
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: sprint-petclinic 
  namespace: argocd-test-demo-2
spec:
  project: default
  source:
    repoURL: https://github.com/redhat-developer/gitops-operator
    path: test/examples/nginx
    targetRevision: HEAD
  destination:
    server: https://kubernetes.default.svc
    namespace: app-ns-2
  syncPolicy:
    automated: {}
----
```

Reviewer: And add the explanations accordingly

Author: Fixed.

+
<1> Specifies the name of an application.
<2> Specifies the namespace used for the first application.
<3> Specifies the name of an application that is created with the same name as the first application.
<4> Specifies the namespace used for the second application.
+
. Next, verify the deployment of each application in the {OCP} Web console.
.. Navigate to *Workloads* → *Pods*.

> .. Ensure that the pods for Argo CD instances, `argocd-instance-1` and `argocd-instance-2` are running.

Author: Fixed.

  namespace: argocd-test-demo-1 # <2>
spec:
  installationID: "instance-demo-1" # <3>
  resourceTrackingMethod: "annotation+label"
----
<1> Specifies the name of the first Argo CD instance.
<2> Specifies the namespace used for the first Argo CD instance.
<3> Specifies the name of the `installationID` for the first Argo CD instance.
+
.Example second Argo CD instance with an annotation label
[source,yaml]
----
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: argocd-instance-demo-2 # <1>
  namespace: argocd-test-demo-2 # <2>
spec:
  installationID: "instance-demo-2" # <3>
  resourceTrackingMethod: "annotation+label"
----
+
<1> Specifies the name of the second Argo CD instance.
<2> Specifies the namespace used for the second Argo CD instance.
<3> Specifies the name of the `installationID` for the second Argo CD instance.
+
. Configure and label target namespaces to associate namespaces with their Argo CD instances.
.. Navigate to *Administration* -> *Namespaces*.
.. Create namespaces for application deployments, `app-ns-1` and `app-ns-2`.
.. To associate each namespace with their respective Argo CD instances, run the following command:
+
[source,terminal]
----

> oc label namespace app-ns-1 argocd.argoproj.io/managed-by=argocd-test-demo-1
> oc label namespace app-ns-2 argocd.argoproj.io/managed-by=argocd-test-demo-2

Reviewer 3: ```suggestion
$ oc label namespace app-ns-1 argocd.argoproj.io/managed-by=argocd-test-demo-1
$ oc label namespace app-ns-2 argocd.argoproj.io/managed-by=argocd-test-demo-2
```

Author: Fixed.

[source,yaml]
----
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: sprint-petclinic # <1>
  namespace: argocd-test-demo-1 # <2>
spec:
  project: default
  source:
    repoURL: https://github.com/redhat-developer/gitops-operator
    path: test/examples/nginx
    targetRevision: HEAD
  destination:
    server: https://kubernetes.default.svc
    namespace: app-ns-1
  syncPolicy:
    automated: {}
----
<1> Specifies the name of the first application.
<2> Specifies the namespace used for the first application.
+
.Example second application using Argo CD
[source,yaml]
----
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: sprint-petclinic # <1>
  namespace: argocd-test-demo-2 # <2>
spec:
  project: default
  source:
    repoURL: https://github.com/redhat-developer/gitops-operator
    path: test/examples/nginx
    targetRevision: HEAD
  destination:
    server: https://kubernetes.default.svc
    namespace: app-ns-2
  syncPolicy:
    automated: {}
----
<1> Specifies the name of the second application that is created with the same name as the first application.
<2> Specifies the namespace used for the second application.

> +

Reviewer 3: ```suggestion

```

Author: Fixed.

. Next, verify the deployment of each application in the {OCP} Web console.
.. Navigate to *Workloads* → *Pods*.
.. Ensure that the pods for Argo CD instances, `argocd-instance-demo-1` and `argocd-instance-demo-2` are running.

> .. Check the application synchronization status in the *Argo CD Applications* tab.

Reviewer: ```suggestion
.. Check the application synchronization status in the *Argo CD Applications* YAML tab or by using the CLI.
```

CLI output example:
```
$ oc get app -n argocd-test-1
NAME        SYNC STATUS   HEALTH STATUS
nginx-app   Synced        Healthy
```

Author: Fixed.

.. Navigate to *Logs* under each Argo CD instance and confirm if you can view the status *Running in namespace-scoped mode*.

> 

Reviewer 3: ```suggestion
+
```

Author: Fixed.

> .. Navigate to *Logs* under each Argo CD instance and confirm if you can view the status *Running in namespace-scoped mode*.

Reviewer: ```suggestion
.. Navigate to argocd-cm configmap in argocd-test-demo-1 and argocd-test-demo-2 namespaces, verify that the installationID is configured successfully.
```

Author: Fixed.

> A successful status message, *Running in namespace-scoped mode*, appears in each Argo CD instance. This status indicates that multiple applications with the same name are successfully deployed in the Argo CD instance.

Reviewer: You can remove this line as there is no such message expected to appear in Argo CD instance

// Module included in the following assemblies:
//
// * argocd_instance/argo-cd-cr-component-properties.adoc

:_mod-docs-content-type: CONCEPT
[id="con_enabling_annotation_based_resource_tracking_in_argo_cd_{context}"]
= Enabling annotation-based resource tracking in Argo CD

{gitops-title} Operator enhances multi-instance support by improving annotation-based resource tracking in Argo CD. Previously, multiple Argo CD instances struggled to differentiate resources with identical application names, relying only on legacy label-based tracking. With this enhancement, users can fully leverage annotation-based tracking by assigning each instance a unique `installationID`, preventing resource conflicts, eliminating infinite sync loops, and enabling safe parallel operation.

> This guide covers setting up multiple Argo CD instances, configuring annotation-based tracking, associating them with namespaces, and verifying deployments by {OCP} Web Console.

Author: Fixed.

[NOTE]
====
* Each Argo CD instance must have a unique `installationID` to prevent resource tracking conflicts.
* Ensure that namespaces are labelled accurately because it allows each Argo CD instance to manage only the intended resources.
* If multiple instances have applications with the same name, set resource tracking to `annotation+label`.

> * If issues arise, check the *Argo CD Application* status and logs in the {OCP} Web Console.

Author: Fixed.

> ====

Author: Fixed.

.Pre-requisites

* You have logged in to the {OCP} cluster as an administrator.
* You have installed the {gitops-title} Operator on your {OCP} cluster.

> [id="configuring-annotation-based-tracking-in-multiple-argo-cd-instances_{context}"]

Author: Fixed.

== Configuring annotation-based tracking in multiple Argo CD instances

> You can configure annotation-based tracking in multiple Argo CD instances by completing the following steps.

Author: Fixed.

<2> Specifies the namespace used for the first Argo CD instance.
<3> Specifies the name of the `installationID` for the first Argo CD instance.
+
.Example second Argo CD instance with an annotation label
[source,yaml]
----
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: argocd-instance-demo-2 # <1>
  namespace: argocd-test-demo-2 # <2>
spec:
  installationID: "instance-demo-2" # <3>
  resourceTrackingMethod: "annotation+label"
----
+
<1> Specifies the name of the second Argo CD instance.
<2> Specifies the namespace used for the second Argo CD instance.
<3> Specifies the name of the `installationID` for the second Argo CD instance.
+
. Configure and label target namespaces to associate namespaces with their Argo CD instances.
.. Navigate to *Administration* -> *Namespaces*.
.. Create namespaces for application deployments, `app-ns-1` and `app-ns-2`.
.. To associate each namespace with their respective Argo CD instances, run the following command:
+
[source,terminal]
----
$ oc label namespace app-ns-1 argocd.argoproj.io/managed-by=argocd-test-demo-1
$ oc label namespace app-ns-2 argocd.argoproj.io/managed-by=argocd-test-demo-2

> ----

Author: Fixed.

+
. Create two applications in Argo CD.

> .. In the OpenShift Web Console, go to *Operators* → *Installed Operators* → *OpenShift GitOps Operator*.

Reviewer 4: ```suggestion
.. In the OpenShift Web Console, go to *Operators* -> *Installed Operators* -> *OpenShift GitOps Operator*.
```

Author: Fixed.

[source,yaml]
----
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: sprint-petclinic # <1>
  namespace: argocd-test-demo-1 # <2>
spec:
  project: default
  source:
    repoURL: https://github.com/redhat-developer/gitops-operator
    path: test/examples/nginx
    targetRevision: HEAD
  destination:
    server: https://kubernetes.default.svc
    namespace: app-ns-1
  syncPolicy:
    automated: {}
----
<1> Specifies the name of the first application.
<2> Specifies the namespace used for the first application.
+
.Example second application using Argo CD
[source,yaml]
----
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: sprint-petclinic # <1>
  namespace: argocd-test-demo-2 # <2>
spec:
  project: default
  source:
    repoURL: https://github.com/redhat-developer/gitops-operator
    path: test/examples/nginx
    targetRevision: HEAD
  destination:
    server: https://kubernetes.default.svc
    namespace: app-ns-2
  syncPolicy:
    automated: {}
----
<1> Specifies the name of the second application that is created with the same name as the first application.
<2> Specifies the namespace used for the second application.

> . Next, verify the deployment of each application in the {OCP} Web console.

Author: Fixed.

> .. Navigate to *Workloads* → *Pods*.

Reviewer 4: ```suggestion
.. Navigate to *Workloads* -> *Pods*.
```

Same suggestion wherever applicable in the PR

Author: Fixed.

.. Ensure that the pods for Argo CD instances, `argocd-instance-demo-1` and `argocd-instance-demo-2` are running.
.. Check the application synchronization status in the *Argo CD Applications* YAML tab or by using the CLI.
+
If you are using the CLI, run the following command to view the output:
+
[source,terminal]
----

> $ oc get app -n argocd-test-1 

Author: Fixed.

----
+
.Example output
----
NAME        SYNC STATUS   HEALTH STATUS

> nginx-app   Synced        Healthy

Reviewer: ```suggestion
sprint-petclinic   Synced        Healthy
```

Author: Fixed.

----

> .. Navigate to the `argocd-cm` configmap in `argocd-test-demo-1` and `argocd-test-demo-2` namespaces and verify that the `installationID` is configured successfully.

Reviewer 4: ```suggestion
.. Navigate to the `argocd-cm` config map in `argocd-test-demo-1` and `argocd-test-demo-2` namespaces and verify that the `installationID` is configured successfully.
```

Author: Fixed.

// Module included in the following assemblies:
//
// * argocd_instance/argo-cd-cr-component-properties.adoc

:_mod-docs-content-type: CONCEPT
[id="con_enabling_annotation_based_resource_tracking_in_argo_cd_{context}"]
= Enabling annotation-based resource tracking in Argo CD

> {gitops-title} Operator enhances multi-instance support by improving annotation-based resource tracking in Argo CD. Previously, multiple Argo CD instances struggled to differentiate resources with identical application names, relying only on legacy label-based tracking. With this enhancement, users can fully leverage annotation-based tracking by assigning each instance a unique `installationID`, preventing resource conflicts, eliminating infinite sync loops, and enabling safe parallel operation.

Author: Fixed.

> This guide covers setting up multiple Argo CD instances, configuring annotation-based tracking, associating them with namespaces, and verifying deployments by {OCP} web console.

Author: Fixed.

[NOTE]
====

> * Each Argo CD instance must have a unique `installationID` to prevent resource tracking conflicts.

Author: Fixed.

// Module included in the following assemblies:
//
// * argocd_instance/argo-cd-cr-component-properties.adoc

:_mod-docs-content-type: CONCEPT
[id="con-enabling-annotation-based-resource-tracking-in-argo-cd_{context}"]
= Enabling annotation-based resource tracking in Argo CD

> {gitops-title} Operator enhances multi-instance support by improving annotation-based resource tracking in Argo CD. Previously, multiple Argo CD instances did not differentiate resources with identical application names properly, relying only on legacy label-based tracking. With this enhancement, you can fully leverage annotation-based tracking by assigning each instance a unique `installationID` ID, preventing resource conflicts, eliminating infinite sync loops, and enabling safe parallel operation.

Author: Fixed.

You can perform the following actions by using the {OCP} web console:
* Set up multiple Argo CD instances
* Configure annotation-based tracking by associating them with namespaces

> * Verify deployments 

Author: Fixed.
